Query refactoring: TermQueryBuilder refactoring and test

cbuescher opened the pull request:
Split the parse(QueryParseContext ctx) method into a parsing and a query building part, adding Streamable for serialization and hashCode(), equals() for better testing.
Add basic unit test for Builder and Parser.

PR goes agains query-refacoring feature branch.

javanna commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -19,22 +19,30 @@
+    private Object value;
 
-    private float boost = -1;
+    float boost = 1.0f;
```
why package private? Other members are private?

cbuescher replied:
changed that in new version of this PR

Commit:
Query refactoring: TermQueryBuilder refactoring and test

Also extended BaseQueryTestCase so it has helper methods for parsing the query header and
extended the toQuery() test method so it passes down parse context to sublass to make
assertions on side effects calling toQuery() has on the parseContext.

cbuescher commented:
Rebased this PR on top of current tip of feature branch and changed test to make use of new BaseQueryTestCase.

javanna commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -147,7 +187,61 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        Query query = null;
+        Preconditions.checkNotNull(this.fieldName, "Fieldname of a TermQuery cannot be null.");
+        Preconditions.checkNotNull(this.value, "Value of a TermQuery cannot be null.");
```
about these checks... we check the same right after parsing (at some point will be moved to coord node), is that correct? and we check again here (which will stay on the data node). I wonder if we should really repeat these checks, or share them between the two methods and introduce some validation method.

cbuescher replied:
I added them here because in theory one could create query objects without parsing them. Maybe a static `boolean validate(QueryBuilder)` on each query builder that also the parsers can use to check before they return a new builder from the fromXContent() method?

cbuescher replied:
I added a validation method for TermQueryBuilder and reused that in the parser. Only downside of this is where previously we had a QueryParseException we now have a ElasticsearchNullPointerException, or we need to try/catch and rethrow which IMHO defeats the purpose. Have a look and let me know what you think.

javanna replied:
First thing which I keep forgetting: you are right, queries can be created without going through `fromXContent`, java api will do just that. There are different places where we need the same validation, I think having a validate method is a good choice.

That said, one question is where do we call it from? `fromXContent` after parsing (on coordinating node in the future) makes perfect sense, but that is called only when the query is provided as json. Java api allows to create the intermediate query representation directly, meaning that the `fromXContent` step will be skipped there.  What can happen in this case is 1) that the query gets serialized to other nodes without being validated, and validated on the data nodes multiple times leading to multiple errors, which is not quite what we want (we want to catch errors once and earlier), 2) that the coordinating node is the data node where the query gets executed, query might not even get serialized, we execute `toQuery` straight-away.

I believe the proper way to validate the query is hooking into the existing request validation mechanism (`ActionRequest#validate`). Every request gets already validated on the coordinating node. The `SearchRequest` should simply call query#validate then I think (only problem being that we cannot quite do it yet, we need to refactor the search request first as right now the whole search request is just a big json object)? 
 
As for exceptions, if we adopt this approach I think we should be consistent with the existing validation infra and have `ActionRequestValidationException`s. 

Does this make sense to you guys?

If so, we need to decide what to do for now, I think having the validate method on the base class makes sense as Lee mentioned. For now we could call it from a single place though (`toQuery` is safer as it covers all the cases), once we move to parsing to the coord node we need to move it to the validation api instead.

dakrone replied:
+1 to this, I think this is a great ide, for now we can do the validate method and figure out where to hook it in at a later time.

Also as an aside, I'd love to get rid of `Preconditions.checkNotNull` and just use `Objects.requireNonNull` since it's part of the JVM, but that's just personal preference.

cbuescher replied:
When I add this to the QueryBuilder interface, I'll have add an empty impl to the BaseQueryBuilder for now (can be removed later once all queries have validate() method). For the details I would like to open a separate issue in which we can then have discussion about exact signature, which exceptions to be thrown, Preconditions vs. Objects.requireNonNull etc... Let me know if you agree with this plan.

cbuescher replied:
Opened #10777 for separate discussion on how to handle validation.

javanna commented on src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -165,4 +172,11 @@ public void testSerialization() throws IOException {
         assertNotSame(deserializedQuery, testQuery);
     }
+
+    private void assertQueryHeader(XContentParser parser) throws IOException {
```
nice! could be static I think

cbuescher replied:
unfortunately not, needs testQuery.parserName() which is not static.

javanna commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -147,7 +187,61 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        MapperService.SmartNameFieldMappers smartNameFieldMappers = parseContext.smartFieldMappers(this.fieldName);
+        if (smartNameFieldMappers != null && smartNameFieldMappers.hasMapper()) {
+            query = smartNameFieldMappers.mapper().termQuery(this.value, parseContext);
+        }
```
can you remind me what happens if the field is not in the mapping? no results and no error? this doesn't change here right?

cbuescher replied:
I have to find out where the mappings are set, hoped you could point me in the right direction. It should be the same code as before in the joint `parse()` method. Will double check.

cbuescher replied:
Code is the same as previously in the Parser. The behaviour when no mapping is found can be controlled by flags in QuerySearchContext. `allowUnmappedFields` flag seems to allow empty (e.g. null) values for the mapper. If thats not true, `mapUnmappedFieldAsString` can be used to get a StringFieldMapper for unmapped fields, else there is an QueryParsingException.

javanna replied:
ok, allowUnmappedField is by default true unless we are parsing query in the percolator or alias filters. Default value can be changed through setting. That means that you can by default query a field that is not mapped I believe.

javanna commented on src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -146,7 +152,8 @@ public void testFromXContent() throws IOException {
     public void testToQuery() throws IOException {
         QueryParseContext context = new QueryParseContext(index, queryParserService);
-        assertLuceneQuery(this.testQuery, this.testQuery.toQuery(context));
+        context.setAllowUnmappedFields(true);
```
interesting... is this supposed to be always true?

cbuescher replied:
I needed to set this for the test to work, because I'm querying random bogus fields. My guess is that the mappings are determined when first docs are indexed, because they have to take the values into account but we don't do this in the unit test setup. Will check this.

cbuescher replied:
see comment above.

rjernst replied:
Can we make query tests not rely on dynamic mappings? allow unmapped fields should not need to be set here.

cbuescher replied:
Would `setMapUnmappedFieldAsString(true)` be okay here? not sure if this is considered "dynamic". Otherwise I get  "Strict field resolution and no field mapping can be found for the field with name [vdFWvqKy]" because I randomly generate fieldnames for this unit test and we don't want to have to create mappings because we dont index anything here.

rjernst replied:
Why are we using randomly generated field names? Can we instead set up mappings at the beginning of the test and select from the fields generated there? I don't think we should be testing *all* of our query tests through unmapped fields; this is a corner case. We should, however, have some tests to check the behavior of unmapped fields with those settings, but the rest of the tests should operate on valid fields.

cbuescher replied:
The fieldnames are random because I mainly want to test the ability of the query parser to deal with varying field names and to check that serialization works. These tests are no integration tests, so I was trying to avoid testing different paths for whether fields have mappings or not. The problem with testing the constructed Lucene queries is that all depends very much on the implementation and testing which check the correct behaviour for mapped/unmapped fields should better be single node / integration tests.

cbuescher replied:
Sorry for confusion, accidentaly deleted the previous comment. Short version was: I was trying to avoid complicated setup for the tests here. If setting mappings for specific fields is easy we should consider it here, otherwise the idea of the base test class so far was to mainly test parsing from Json and to check that serialization and query construction works. Checking that the constructed lucene queries really work on an index might be better in a single node / integration test or what do you think?

javanna replied:
would love to keep this a unit test that doesn't require nodes, that said I am not sure what we have to do to make it work by defining mappings etc. seems like we might have to move to single node test.... WDYT Ryan?

rjernst replied:
I had thought this was a singlenodetest, my mistake! But it does look like you are creating an IndexQueryParserService, which has a MapperService, yet I don't see it in the initialized injection? 

That said, you can fake this out with a test QueryParserContext? That is what each query parser calls to get the mapping for a field. Otherwise always `unmappedFieldAsString` you would always get a string field, but this won't work for other query parsers right?

cbuescher replied:
For the TermQueryBuilder that is tested in this PR there are two paths in `toQuery(context)` depending on whether `smartFieldMappers` are found or not in the context. I think in this case we don't need to test all possible cases since it's hard to test all the internals of the produced lucene query without getting to implementation specific (doing casts, instanceof etc...) so I would opt for using `unmappedFieldAsString` for simplicity here. 
However, when we see writing tests for other queries that we need to test different execution paths depending on whether there is a mapping or not, we can fake other QueryParseContexts there and overwrite the BaseQueryTestCase#testToQuery() case there.

javanna replied:
I am conflicted here: on one hand I don't like that we cover only one codepath (when the field doesn't exist), on the other hand I don't want to be testing mappers here too and how they generate queries. I think we can go ahead with unmapped fields at least for now.

javanna commented on src/test/java/org/elasticsearch/index/query/TermQueryBuilderTest.java:
```diff
@@ -0,0 +1,106 @@
+            case 1: value = randomAsciiOfLength(8); break;
+            case 2: value = randomInt(10000); break;
+            case 3: value = randomDouble(); break;
+            default: value = randomAsciiOfLength(8);
```
you don't need the default do you?

cbuescher replied:
Yep.

javanna commented on src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -57,12 +58,14 @@
-    private static Index index;
+    protected static Injector injector;
+    protected static IndexQueryParserService queryParserService;
+    protected static Index index;
```
do we need to make everything protected? how about exposing a protected method that creates the QueryParseContext which you need in the additional tests?

javanna commented on src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -116,8 +119,9 @@ public static void after() throws Exception {
 
     /**
      * Subclass should handle assertions on the lucene query produced by the query builder under test here
+     * @param context
```
description please, or remove the param line

javanna commented on src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -116,8 +119,9 @@ public static void after() throws Exception {
+     * @param context
      */
-    protected abstract void assertLuceneQuery(QB queryBuilder, Query query) throws IOException;
+    protected abstract void assertLuceneQuery(QB queryBuilder, Query query, QueryParseContext context) throws IOException;
```
the more I see QueryParseContext used in toQuery the more I think we should start splitting it in two objects, one for parsing and one for toQuery only. But maybe I am jumping ahead a bit, we can in the meantime familiarize with what it is used for and once we migrated all the queries go and do that other step? What do you think?

cbuescher replied:
Yes, I would first try to keep track which queries use it for which purpose.

javanna replied:
I agree on not making this change here.

javanna commented on src/test/java/org/elasticsearch/index/query/TermQueryBuilderTest.java:
```diff
@@ -0,0 +1,106 @@
+            } else {
+                fail("Expected either a QueryWrapperFilter or a CustomQueryWrappingFilter to be registered under " + queryBuilder.queryName() +
+                        " in the query parse context");
+            }
```
I wish there was a less impl dependent way of checking that named queries/filters get parsed (without instanceof etc.). Looks good though I don't see many other ways to do this. Btw this is code that every single query should support I believe (apart from match_all?), maybe we should try and make it more generic if possible (not only test wise).

cbuescher replied:
Agreed, but I would prefer looking at that in a separate issue, because making this lookup more generic will probably touch more queries.

cbuescher replied:
I just checked, from the 37 queries currently implementing QueryBuilder only 7 don't support having a "queryName":
  * BoostingQB
  * ConstantScoreQB
  * FunctionScoreQB
  * MatchAllQB
  * SpanMultiTermQB
  * TemplateQB
  * WrapperQB

However, making this more generic should probably be a separate issue.

javanna replied:
different question...wonder if the above queries should support query name too then....we should improve this, agreed it's a different PR. making it more generic might get tricky parse wise, we'll see. The test stll seems a bit too low level with those casts but I have no good ideas on how to improve it to be honest.

javanna replied:
let's not make this hold this PR, but let's keep track of this potential issue and address the need for generifying in a separate issue

cbuescher replied:
I opened https://github.com/elastic/elasticsearch/issues/10776 for keeping track of ideas to generify the handling of the queryName field.

javanna commented:
left a few comments

cbuescher commented:
I pulled out the validation of the two important fields into a method shared by builder and parser part, but not sure if this is the best way to go. Happy about comments there. Hope I adressed the rest of your comments, I would prefer tracking the use of QueryParseContext in toQuery() in a separate issue before making huge changes there. Same goes for generifying named queries (lookup but also maybe simplification)

dakrone commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -105,10 +113,28 @@ public TermQueryBuilder(String name, boolean value) {
      */
     public TermQueryBuilder(String name, Object value) {
-        this.name = name;
+        this.fieldName = name;
```
Can you change `name` to `fieldName` in the constructor too?

cbuescher replied:
done

dakrone commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -147,7 +187,65 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        return query;
+    }
+
+    protected void validate() {
```
Are all the refactored queries going to have something like a `validate` method? I wonder if it makes sense to add it to the superclass for them all too, so it's available on any query.

javanna replied:
+1, see also https://github.com/elastic/elasticsearch/pull/10669#discussion_r29000261 @dakrone would love to hear if it makes sense to you ;)

cbuescher replied:
I added `public void validate()` to QueryBuilder now and added empty impl to be removed later to `BaseQueryBuilder`. Correct me if this is not what you intended.

cbuescher commented:
Changed variable name in constructor and added validate() to the QueryBuilder interface. As long as this is not implemented by all queries, added empty impl to BaseQueryBuilder. Also opened two separate issues to keep track of further ideas for validation (#10777) and for generifying named queries (#10776). 
I'm still not sure about how to improve the tests for `toQuery()` because the resulting lucene queries hide a lot of details, so assertions depend on explicit casts and knowing implementation details that might break often. I'm inclined to leave the `toQuery()`tests here very general and rely on other integration tests with real cluster setup / mappings etc. to check that queries work. Thoughts on this appreciated.

dakrone commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -147,7 +187,65 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        return query;
+    }
+
+    public void validate() {
```
Maybe add the `@Override` annotation here?

cbuescher commented on src/main/java/org/elasticsearch/index/query/TermQueryParser.java:
```diff
@@ -80,27 +74,19 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
             parser.nextToken();
         } else {
-            value = parser.text();
+            value = parser.objectText();
```
One question I had myself here: I had to change this for the `fromXContentent()" test to work. Otherwise, if the value is e.g. an int, it is written to the XContent query as a number value. With the original `parser.text()` it is then read in as String (which doesn't seem to matter for the later toQuery(), since there BytesRefs.toBytesRef(value) is used, which calls toString() anyway) but blows up the newly introduced equality test. I changes this to `objectText()` because this does some checking based in the JsonToken type.

cbuescher replied:
Thinking about it, I tend to want to revert this change since I'm not sure about it's implications. The problem then is that `Object value` can be of different type in original query and after `fromXContent()`, but in the end it only matters that `toQuery()` produces the same result, and there every Object is converted to lucene BytesRef. Thats why I tend to use that conversion also in the `TermQueryBuilder#equals()`. Will push that change to show what I mean.

javanna replied:
we discussed this and we went for always keeping `BytesRef` within `TermQueryBuilder` rather than mixing up `String` and `BytesRef` depending on whether we received the query via java api or we parsed it through json. Java api will still have a setter that accepts a string, but internally that setter will convert from `String` to `BytesRef`.

We do have to fix the above `parser.text()` which looks like a bug that never manifested as when parsing and executing  on the same (data) node everything still works. But we are currenly parsing the value in the term query short format (`term: { field: value}`) always as a string.

cbuescher commented on src/main/java/org/elasticsearch/index/query/TermQueryParser.java:
```diff
@@ -66,9 +60,9 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
                 } else {
                     if ("term".equals(currentFieldName)) {
-                        value = parser.objectBytes();
+                        value = parser.objectText();
```
Also would like feedback on this change: Did this for `fromXContent()` tests to work. When value field is String, was read in here as BytesRef (internally call to JsonXContentparser#utf8Bytes()). This would lead to non-equal queries for tests, original value being String, after `fromXContent` BytesRef. Again, this shouldn't matter once lucene query is constructed. I changed to `objectText()` here, alternatively I could convert back for equality test to work, but that seems awkward.
```
value = parser.objectBytes();
if (value instanceof BytesRef) {
   value = ((BytesRef) value).utf8ToString();
}
```
Thoughts?

cbuescher replied:
See comment above.

cbuescher commented:
@javanna went through the comments and current diff again, here are the things that are open from my point of view:
* using `context.setMapUnmappedFieldAsString(true)` in the BaseQueryTestCase
* using `BytesRefs.toBytesRef(value)` in `equals()` and `hashCode()` from my last update commit

Anything else open here that I'm missing?

javanna commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -147,7 +188,66 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+    @Override
+    public void validate() throws ElasticsearchNullPointerException {
+        Preconditions.checkNotNull(this.fieldName, "Fieldname of a TermQuery cannot be null.");
+        Preconditions.checkNotNull(this.value, "Value of a TermQuery cannot be null.");
```
I think we have to check if these are empty too and barf in that case?

javanna replied:
one more thing: s/Fieldname/ field name
I think in general the first line of an error should be lowercase, that's pretty much the convention we use in the existing validation code.

javanna commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -120,22 +147,36 @@ public TermQueryBuilder boost(float boost) {
+    /**
+     * Gets the query name for the query.
+     */
+    public String queryName() {
```
at this point I 'd probably move the whole queryName code to the base class including serialization bits too. Up to you if you want to do it here or in a separate PR

cbuescher replied:
I'd probably try doing this for all queries in #10776, started looking at this on separate branch already. There's a few more things to clean here, so I'd probably not mix it but revisit this query again in separate PR.

javanna commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -147,7 +188,66 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
 
     @Override
+    public Query toQuery(QueryParseContext parseContext) {
+        validate();
```
I think I would just call validate after parsing for now, which will be replaced in a later stage by the validate request infra. Makes little sense to call it from here to be honest. I would remove.

javanna commented on src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java:
```diff
@@ -79,4 +79,10 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
+    public void validate() {
+        // default impl does not validate, subclasses should override.
+        // TODO Remove once all queries support validation
+    };
```
I don't think we need this if we move over query by query to extend this base class?

javanna replied:
sorry I keep getting confused between parser and builder, we do need it. can you remove the `;` at the end?

javanna commented on src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java:
```diff
@@ -147,7 +188,66 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(fieldName, BytesRefs.toBytesRef(value), boost, queryName);
```
not using bytesRef here would have the advantage of keeping `equals` and `hashcode` independent from lucene, which kinda sounds good to me.

cbuescher replied:
Agreed, but this brings me back to my previous problem mentioned in this previous comment https://github.com/elastic/elasticsearch/pull/10669#discussion-diff-29134022. I don't know if it is save parsing the query value using `parser.objectText()` instead of `parser.objectBytes()`, but otherwise the simple equals doesn't work.

javanna commented on src/main/java/org/elasticsearch/index/query/QueryBuilder.java:
```diff
@@ -45,4 +45,10 @@
+     * Validate the query.
+     * @throws ElasticsearchException if the query is not valid, e.g. fields that are needed to create the lucene query are missing.
+     */
+    void validate() throws ElasticsearchException;
```
as for the signature here, we might not just want to throw exception at the first error encountered, but rather collect them all. That would mean returning a list of errors, or a single exception that wraps multiple validation errors as we do in ActionRequest#validate. Thoughts on this?

cbuescher replied:
Looked at ActionRequestValidationException, could make a copy of that as QueryValidationException. It seems like they are never thrown though, rather returned, so signature would be `QueryValidationException validate()` then? Could also collect all errors and then throw exception at the end.

javanna replied:
I think we should stick to the current validate api for now, although it does feel weird that we don't actually throw exception. What happens is that the listener gets notified of the exception in `TransportAction#execute`. We can't yet hook into the validate mechanism as the `SearchRequest` only contains a `source` bytes array in json format, but at the end of the refactoring we should have different elements that hold the different parts of a search request, among which the query which can be validated as part of `SearchRequest#validate`. This is to me the only way to make sure that validation happens whenever needed.

cbuescher replied:
So you suggest I change signature to `ActionRequestValidationException validate()` and don't throw exceptions, but instead add meaningful error messages to that exception, right? Will change that then, can't use Preconditions then because they throw RuntimeExceptions, but thats not a big problem.

javanna replied:
indeed I suggest to move away from `Preconditions`, for now using ActionRequestValidationException sounds good, maybe a specific query exception would be better but we can figure it out later on I guess

cbuescher commented:
I added the changes to the validate() signature as discussed. Also changed internal representation of the query value in case it is a String to BytesRef to be compatible with parser.objectBytes() on the parser side. If this looks okay to you, I can also merge the additions to StreamInput/Output to master in a separate PR already.

cbuescher commented:
Added one more change addressing the merging of two validation cases.

javanna commented on src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -388,6 +388,9 @@ public void writeGenericValue(@Nullable Object value) throws IOException {
             writeDoubleArray((double[]) value);
+        } else if (type == BytesRef.class) {
+            writeByte((byte) 21);
+            writeBytesRef((BytesRef) value);
```
if you rebase you'll get this change that was made upstream and merged back into our branch

javanna commented on src/main/java/org/elasticsearch/index/query/QueryBuilder.java:
```diff
@@ -45,4 +46,11 @@
+     * @return {@code null} if query is valid, otherwise {@link ActionRequestValidationException} containing error messages,
+     * e.g. if fields that are needed to create the lucene query are missing.
+     */
+    ActionRequestValidationException validate();
```
sorry for changing my mind, what do you think about having our own exception here e.g. `QueryValidationException` ? when we'll hook into the search request validation we can just wrap it in the ActionRequestValidationException or something along those lines. I wouldn't query builders to depend on transport action packages ideally.

cbuescher replied:
Changed the method signature to `QueryValidationException validate()` and introduced own exception class, similar to the existing one.

javanna commented on src/test/java/org/elasticsearch/index/query/TermQueryBuilderTest.java:
```diff
@@ -0,0 +1,114 @@
+            }
+        }
+    }
+
```
should we have a test (possibly in the base class) that also verifies for each query what happens when we parse invalid json?

cbuescher replied:
I think that's a good idea. However, this seems to be difficult in a generic way because there are many ways json can be invalid and the test setup needs to take implementation details into account (which exception will be thrown, which fieldnames are allowed. I tried this in the base class but ended up having to manipulate the query string in very crude ways. I'm going to open an issue for this so we can collect and discuss ideas on this. I think it doesn't have to go in with this PR.

cbuescher replied:
Opened #10974 for tracking this.

javanna commented on src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java:
```diff
@@ -79,4 +80,11 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
+    @Override
+    public ActionRequestValidationException validate() {
+        // default impl does not validate, subclasses should override.
+        // TODO Remove once all queries support validation
```
can you use //norelease instead of TODO? then we are sure we don't release if this method is there cause the build script stops the build in that case

cbuescher replied:
done

Commit:
Add //norelease to validation method in BaseQueryBuilder

Commit:
Added QueryValidationException to have own exception type for queries

Commit:
Added 20 iteration repeat to test

cbuescher commented:
Rebased this PR on current feature branch and changed validation exception class to own version similar to the previously suggested ActionRequestValidationException. Also opened separate issue #10974 for keeping track of adding tests for invalid json and set the TermQueryBuilder test to 20 repetitions.

javanna commented on src/main/java/org/elasticsearch/index/query/QueryBuilder.java:
```diff
@@ -44,4 +45,11 @@
+
+    /**
+     * Validate the query.
+     * @return {@code null} if query is valid, otherwise {@link ActionRequestValidationException} containing error messages,
```
need to update exception in javadocs

javanna commented on src/main/java/org/elasticsearch/index/query/QueryValidationException.java:
```diff
@@ -0,0 +1,77 @@
+
+    private final List<String> validationErrors = new ArrayList<>();
+
+    public QueryValidationException() {
```
maybe make the first validation error mandatory, by using a constructor that accepts an error rather than the default one?

javanna commented on src/main/java/org/elasticsearch/index/query/QueryValidationException.java:
```diff
@@ -0,0 +1,77 @@
+
+    /**
+     * Helper method than can be used to add error messages to an existing {@link QueryValidationException}.
+     * When passing @null as the initial exception, a new exception is created.
```
is `@null` a valid javadocs tag?

javanna commented on src/test/java/org/elasticsearch/index/query/TermQueryBuilderTest.java:
```diff
@@ -0,0 +1,117 @@
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+
+@Repeat(iterations=20)
```
this is fine for now, but I'd rather have more runs on CI than spreading these repeats all over the place. We are most likely going to forget about them, already too many things we have to watch out for in this branch.

javanna commented:
LGTM besides the minor comments left, if you can address them that would be great, this is good to merge then!

Commit:
Minor changes adressing last comments

cbuescher commented:
Thanks for the quick response, pushed this to the feature branch.

cbuescher commented:
Thanks for the review and for closing the issue, somehow it got opened
again yesterday automatically.

On Wed, May 6, 2015 at 7:34 AM, Luca Cavanna <notifications@github.com>
wrote:

> Closed #10669 <https://github.com/elastic/elasticsearch/pull/10669>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/10669#event-297897046>.
>



-- 
Christoph Büscher
